core/metadata.cpp

```diff
@@ -0,0 +1,192 @@
+// SPDX-License-Identifier: GPL-2.0
+#include "metadata.h"
+#include "exif.h"
+#include "qthelper.h"
+#include <QString>
+#include <QFile>
+#include <QDateTime>
+
+// Fetch unsigned integer of arbitrary length from file or memory in big-endian order.
+// On file error, return default value.
+// This is hackish, but it works.
+// Note: only works for unsigned integer types. This is not tested for!
+template <typename T>
+static inline T getBE(const char *buf_in)
+{
+	constexpr size_t size = sizeof(T);
+	auto buf = reinterpret_cast<unsigned const char *>(buf_in);
+	T ret = 0;
+	for (size_t i = 0; i < size; ++i)
+		ret = (ret << 8) | buf[i];
+	return ret;
+}
```

Comment:
personally i don't mind this that much.
templates are OK as long as they don't infest the code base too much.

i think one way to avoid the template usage here is to always return the largest possible integer type:
```
static inline uint64_t get_be(const char *buf_in, size_t size)
{
	unsigned const char *buf = (unsigned const char *)buf_in;
	uint64_t ret = 0;
	for (size_t i = 0; i < size; ++i)
		ret = (ret << 8) | buf[i];
	return ret;
}
```
this maybe has higher inline chances but drops the compile time calculations and is less safe.
it might also get slower for small `size` demands, depending on how the compiler deals with `uint64_t` on a 32bit register machine.

Reply:
Sure, that would be possible as well. But promoting everything to 64 bit just to avoid the template-syntax seems a bit strange and hardly more elegant. And what if one day we want to extract 128 bit values?

I'm travelling the next two days and will then clean things up with more/better comments and unify the `quint32` vs. `uint32_t` mess. Hopefully, this will make the code less hard to read.

Reply:
there is the C way of doing N bit values too (if there is no built in N bit type), but templates become marginally more easy to use for this.